Implemented Introsort

Kishan-Ved opened the pull request:
#### References to other Issues or PRs or Relevant literature
Fixes https://github.com/codezonediitj/pydatastructs/issues/545

#### Brief description of what is fixed or changed
Implemented Introsort.

czgdp1807 commented on pydatastructs/linear_data_structures/tests/benchmarks/test_algorithms.py:
```diff
@@ -41,6 +41,9 @@ def test_quick_sort():
 def test_bubble_sort():
     _test_common_sort(bubble_sort, size=2000)
 
+# @pytest.mark.xfail
+# def test_intro_sort():
+#     _test_common_sort(intro_sort, size=2000)
```
```suggestion
@pytest.mark.xfail
def test_intro_sort():
     _test_common_sort(intro_sort, size=2000)
```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,113 @@ def jump_search(array, value, **kwargs):
+        maxdepth = 0
+    else:
+        maxdepth = kwargs.get("maxdepth", int(2 * (log(n)/log(2))))
+    ins_threshold = kwargs.get("ins_threshold",16)
```
```suggestion

    ins_threshold = kwargs.get("ins_threshold", 16)
```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,113 @@ def jump_search(array, value, **kwargs):
+
+def intro_sort(array, **kwargs) -> Array:
+    """
+    Performs quick sort on the given array.
```
```suggestion
    Performs intro sort on the given array.
```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,113 @@ def jump_search(array, value, **kwargs):
+    """
+    backend = kwargs.pop("backend", Backend.PYTHON)
+    if backend == Backend.CPP:
+        return _algorithms.intro_sort(array, **kwargs)
```
```suggestion
        return _algorithms.intro_sort(array, **kwargs)

```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,113 @@ def jump_search(array, value, **kwargs):
+        array[lower], array[right] = array[right], array[lower]
+        return right
+
+    from pydatastructs.trees.heaps import BinaryHeap
```
```suggestion
    from pydatastructs.trees.heaps import BinaryHeap

```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,113 @@ def jump_search(array, value, **kwargs):
+        return right
+
+    from pydatastructs.trees.heaps import BinaryHeap
+    if(n<ins_threshold):
+        return insertion_sort(array, start=lower,end=upper)
+
+    elif(maxdepth==0):
+        heapsort(array,start=lower,end=upper)
+        return array
+
+    else:
+        p = partition(array,lower,upper)
```
```suggestion
    if n < ins_threshold:
        return insertion_sort(array, start=lower, end=upper)
    elif maxdepth == 0:
        heapsort(array, start=lower, end=upper)
        return array
    else:
        p = partition(array, lower, upper)
```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,113 @@ def jump_search(array, value, **kwargs):
+    else:
+        p = partition(array,lower,upper)
+
+        intro_sort(array, start=lower, end=p-1, maxdepth=maxdepth-1,ins_threshold=ins_threshold)
+        intro_sort(array, start=p+1, end=upper,  maxdepth=maxdepth-1,ins_threshold=ins_threshold)
```
```suggestion
        intro_sort(array, start=lower, end=p-1, maxdepth=maxdepth-1, ins_threshold=ins_threshold)
        intro_sort(array, start=p+1, end=upper,  maxdepth=maxdepth-1, ins_threshold=ins_threshold)
```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,114 @@ def jump_search(array, value, **kwargs):
+        return _algorithms.intro_sort(array, **kwargs)
+
+    # from pydatastructs import Stack
+    # Always sorts in increasing order
```
I think this is because of heapsort's limitation, right? Please mention here about this.

Kishan-Ved replied:
Yes, that is because of the current heapsort function's limitation.

Kishan-Ved replied:
Should I write that as a comment in the code?

Kishan-Ved replied:
Changes made.

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,114 @@ def jump_search(array, value, **kwargs):
+    if backend == Backend.CPP:
+        return _algorithms.intro_sort(array, **kwargs)
+
+    # from pydatastructs import Stack
```
I think this isn't needed?

Kishan-Ved replied:
Yes import stack is not needed.

Kishan-Ved replied:
For backend.CPP, I thought someone might write that code in future, so I kept it.

Kishan-Ved replied:
I will implement any further changes/suggestions tomorrow morning. Thank you!

Kishan-Ved replied:
Changes made.

Kishan-Ved commented:
I've done the mentioned changes: Added comment to code that a custom compactor cannot be taken due to heapsort's limitation.
I've also removed the commented statement of import stack.

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,115 @@ def jump_search(array, value, **kwargs):
+        left = lower + 1
+        right = upper
+        done = False
+        while not done:
+            while left <= right and _comp(array[left],pivot,comp):
+                left += 1
+            while _comp(pivot,array[right],comp) and right >= left:
+                right -= 1
+            if right < left:
+                done = True
+            else:
+                array[left], array[right] = array[right], array[left]
+                left+=1
+                right-=1
+
+        array[lower], array[right] = array[right], array[lower]
+        return right
+
```
```suggestion
        while not done:
            while left <= right and _comp(array[left], pivot, comp):
                left += 1
            while _comp(pivot, array[right], comp) and right >= left:
                right -= 1
            if right < left:
                done = True
            else:
                array[left], array[right] = array[right], array[left]
                left+=1
                right-=1

        array[lower], array[right] = array[right], array[lower]
        return right

```

czgdp1807 commented on pydatastructs/linear_data_structures/algorithms.py:
```diff
@@ -1740,3 +1741,115 @@ def jump_search(array, value, **kwargs):
+
+    .. [1] https://en.wikipedia.org/wiki/Introsort
+    """
+    backend = kwargs.pop("backend", Backend.PYTHON)
+    if backend == Backend.CPP:
+        return _algorithms.intro_sort(array, **kwargs)
```
```suggestion
    raise_if_backend_is_not_python(
        next_permutation, kwargs.get('backend', Backend.PYTHON))
```

Kishan-Ved commented:
1 check didn't pass, I believe this is due to some backend issue.
```
=========================== short test summary info ============================
FAILED pydatastructs/utils/tests/test_code_quality.py::test_backend_argument_message - AssertionError: assert '<function intro_sort at 0x7f97b968adc0>' in 'As of 1.0.1-dev version, only Python backend is supported for <function next_permutation at 0x7f97b968a940> API'
 +  where '<function intro_sort at 0x7f97b968adc0>' = str(<function intro_sort at 0x7f97b968adc0>)
======= 1 failed, 160 passed, 2 xfailed, 7 xpassed in 138.76s (0:02:18) ========
```

Kishan-Ved commented:
All checks pass, the code suggested had next_permutation by mistake, which I changed to my function: intro_sort

Kishan-Ved commented:
Thank you for your help! @czgdp1807 
This is one of my first prominent open source code contributions! This helped me learn about reading, understanding and managing a large codebase. I'm delighted and eager to contribute more to this organization :)

Also please check the issue I faced, which I've mentioned here: https://github.com/codezonediitj/pydatastructs/issues/547 , along with the solution.
